# Sensor Params
sensor_params:
  imuRate: 400                             #Rate of IMU input (Hz)
  imuBufferLength: 800
  lidarOdometryRate: 5
  gnssRate: 10
  imuTimeOffset: 0.0 # Offset between IMU and LiDAR Measurements: can be determined with rqt_multiplot

# Factor Graph
graph_params:
  useIsam: true # if false, then levenberg-marquardt is used
  smootherLag: 0.5                         #Lag of fixed lag smoother[seconds], not needed for ISAM2
  additionalOptimizationIterations: 0                  #Additional iterations of graph optimizer after update with new factors
  findUnusedFactorSlots: true
  enableDetailedResults: false
  usingFallbackGraph: false
  usingCholeskyFactorization: true # CHOLESKY faster, QR numerically more stable

# Outlier Rejection
outlier_params:
  gnssOutlierThreshold: 0.2 # in meters, if jumping more than this, it is considered as absent GNSS, occurs between two GNSS measurements

# Noise Parameters
noise_params:
  ## IMU
  ### Accelerometer
  accNoiseDensity: 7.84e-04                 #Continuous-time "Covariance" of accelerometer, microstrain: sigma^2=7.225e-7
  accBiasRandomWalk: 1.0e-03                #Continuous-time "Covariance" describing accelerometer bias random walk, default: 1.0e-06
  accBiasPrior: 0.0                         #Prior/starting value of accelerometer bias
  ### Gyro
  gyrNoiseDensity: 3.4906585e-05               #Continuous-time "Covariance" of gyroscope measurements, microstrain: sigma^2=1.71-08
  gyrBiasRandomWalk: 1.0e-04 #gnss:9.33e-08 #lidar: 9.33e-01            #Continuous-time "Covariance" describing gyroscope bias random walk, default: 9.33e-08
  gyrBiasPrior: 0.0                        #Prior/starting value of gyroscope bias
  ### Preintegration
  integrationNoiseDensity: 1.0e-03         #continuous-time "Covariance" describing integration uncertainty, default: 1.0e-06
  biasAccOmegaPreInt: 1.0e-02             #covariance of bias used for preintegration, default: 1.0e-2
  ## LiDAR
  poseBetweenNoise: [ 5.0, 5.0, 5.0, 2.0, 2.0, 2.0 ] # gnss [ 10.0, 10.0, 10.0, 2.0, 2.0, 2.0 ] # lidar: [ 1000.0, 1000.0, 1000.0, 100.0, 100.0, 100.0 ] #Noise of add between factor -ORDER RPY(rad) - XYZ(meters)
  poseUnaryNoise: [ 10.0, 10.0, 1.0, 5.0, 5.0, 5.0 ] # ORDER RPY(rad) - XYZ(meters) First tests:  [ 1000.0, 1000.0, 1000.0, 2.0, 2.0, 2.0 ]
  ## GNSS
  gnssPositionUnaryNoise: 1.0 #1.0e-02 # x, y, z of global position
  gnssHeadingUnaryNoise: 1.0 # x,y,z of heading vector

# Relinearization
relinearization_params:
  positionReLinTh: 0.005                    #Position linearization threshold
  rotationReLinTh: 0.005                    #Rotation linearization threshold
  velocityReLinTh: 0.001                     #Linear Velocity linearization threshold
  accBiasReLinTh: 0.001                      #Accelerometer bias linearization threshold
  gyrBiasReLinTh: 0.001                      #Gyroscope bias linearization threshold
  relinearizeSkip: 0
  enableRelinearization: true
  evaluateNonlinearError: false
  cacheLinearizedFactors: true
  enablePartialRelinearizationCheck: false
